[Headers][X86] VisitCallExpr constexpr immediate shifts (#154293)

Arghnews opened the pull request:
Fixes #154293 

Implement VectorExprEvaluator::VisitCallExpr constexpr support for left, right, arithmetic shift for MMX/SSE/AVX2/AVX512 intrinsics

_mm*_slli_epi*
_mm*_srli_epi*
_mm*_srai_epi*
_mm*_mask_slli_epi*
_mm*_maskz_slli_epi*

NOTE: not all intrinsics have all widths i.e.
_mm_srli_pi32 doesn't have pi64 etc.

Let me know any other feedback as it stands, thanks

Implements (full list):
```
_mm256_mask_slli_epi16
_mm256_mask_slli_epi32
_mm256_mask_slli_epi64
_mm256_mask_srai_epi16
_mm256_mask_srai_epi32
_mm256_mask_srai_epi64
_mm256_mask_srli_epi16
_mm256_mask_srli_epi32
_mm256_mask_srli_epi64
_mm256_maskz_slli_epi16
_mm256_maskz_slli_epi32
_mm256_maskz_slli_epi64
_mm256_maskz_srai_epi16
_mm256_maskz_srai_epi32
_mm256_maskz_srai_epi64
_mm256_maskz_srli_epi16
_mm256_maskz_srli_epi32
_mm256_maskz_srli_epi64
_mm256_slli_epi16
_mm256_slli_epi32
_mm256_slli_epi64
_mm256_srai_epi16
_mm256_srai_epi32
_mm256_srai_epi64
_mm256_srli_epi16
_mm256_srli_epi32
_mm256_srli_epi64
_mm512_mask_slli_epi16
_mm512_mask_slli_epi32
_mm512_mask_slli_epi64
_mm512_mask_srai_epi16
_mm512_mask_srai_epi32
_mm512_mask_srai_epi64
_mm512_mask_srli_epi16
_mm512_mask_srli_epi32
_mm512_mask_srli_epi64
_mm512_maskz_slli_epi16
_mm512_maskz_slli_epi32
_mm512_maskz_slli_epi64
_mm512_maskz_srai_epi16
_mm512_maskz_srai_epi32
_mm512_maskz_srai_epi64
_mm512_maskz_srli_epi16
_mm512_maskz_srli_epi32
_mm512_maskz_srli_epi64
_mm512_slli_epi16
_mm512_slli_epi32
_mm512_slli_epi64
_mm512_srai_epi16
_mm512_srai_epi32
_mm512_srai_epi64
_mm512_srli_epi16
_mm512_srli_epi32
_mm512_srli_epi64
_mm_slli_epi16
_mm_slli_epi32
_mm_slli_epi64
_mm_slli_pi16
_mm_slli_pi32
_mm_srai_epi16
_mm_srai_epi32
_mm_srai_epi64
_mm_srai_pi16
_mm_srai_pi32
_mm_srli_epi16
_mm_srli_epi32
_mm_srli_epi64
_mm_srli_pi16
```

Arghnews commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -11644,64 +11679,181 @@ bool VectorExprEvaluator::VisitCallExpr(const CallExpr *E) {
+            std::move(ResultOut),
+            /*isUnsigned=*/DestEltTy->isUnsignedIntegerOrEnumerationType())));
+      } else {
+        APSInt RHS = SourceRHS.getVectorElt(EltNum).getInt();
```
For reviewer benefit, I'll point out this `else` case and all of it below, I didn't touch, except to insert this `if` and my change above. I suspect this is the result of rebase-ing onto latest and a change coming in here that I had to fix the merge conflict from

RKSimon replied:
You can reduce the diff and avoid the else case if you continue at the end of the SourceRHS.isInt() block

RKSimon commented:
CC @tbaederr who can advise on helping you add support for the new interp as well

RKSimon commented:
@Arghnews Maybe use #155358 as reference

RKSimon commented on clang/lib/AST/ByteCode/Pointer.h:
```diff
@@ -72,7 +72,7 @@ enum class Storage { Block, Int, Fn, Typeid };
 ///
 /// The Offset field is used to access the actual data. In other words, the
-/// data the pointer decribes can be found at
+/// data the pointer describes can be found at
```
irrelevant - if you want to fix this - handle it in a separate patch

RKSimon commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -11646,64 +11677,150 @@ bool VectorExprEvaluator::VisitCallExpr(const CallExpr *E) {
+        case clang::X86::BI__builtin_ia32_psraqi256:
+        case clang::X86::BI__builtin_ia32_psraqi512:
+          ResultElements.push_back(APValue(APSInt(
+              LHS.ashr(std::min(ShiftAmount, LaneBitWidth)), DestUnsigned)));
```
LaneBitWidth - 1 ?

Arghnews commented:
Have addressed feedback @RKSimon

Will wait on https://github.com/llvm/llvm-project/pull/155891 then I can refactor this based on what goes in (and squash)

Let me know if I should do otherwise or anything else needs doing on this, thanks

RKSimon commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -11704,6 +11793,33 @@ bool VectorExprEvaluator::VisitCallExpr(const CallExpr *E) {
         ResultElements.push_back(
             APValue(APSInt(LHS.lshr(RHS.getZExtValue()), DestUnsigned)));
         break;
+        APSInt RHS = SourceRHS.getVectorElt(EltNum).getInt();
```
This looks like somethings gone wrong in the merge?

RKSimon commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -11704,6 +11793,32 @@ bool VectorExprEvaluator::VisitCallExpr(const CallExpr *E) {
+              APValue(APSInt(llvm::APIntOps::mulhs(LHS, RHS),
+                             /*isUnsigned=*/false)));
+          break;
+        }
```
this block needs removing entirely

Arghnews commented:
Refactored to use your @RKSimon WIP change in #156017 , much neater way of doing it, cheers

RKSimon commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -11646,6 +11677,64 @@ bool VectorExprEvaluator::VisitCallExpr(const CallExpr *E) {
+        case clang::X86::BI__builtin_ia32_psllqi128:
+        case clang::X86::BI__builtin_ia32_psllqi256:
+        case clang::X86::BI__builtin_ia32_psllqi512:
+          if (ShiftAmount >= LHS.getBitWidth()) {
```
LHS,getBitWidth() -> LaneBitWidth

RKSimon reviewed (commented):
LGTM - but let @tbaederr comment on the bytecode

Arghnews committed:
[Headers][X86] VisitCallExpr constexpr immediate shifts (#154293)

Implement VectorExprEvaluator::VisitCallExpr constexpr support for left,
right, arithmetic shift for MMX/SSE/AVX2/AVX512 intrinsics.
Also implement in experimental-new-constant-interpreter

Adds support and tests for
_mm*_slli_epi*
_mm*_srli_epi*
_mm*_srai_epi*
_mm*_mask_slli_epi*
_mm*_maskz_slli_epi*

NOTE: not all intrinsics have all widths i.e.
_mm_srli_pi32 doesn't have pi64 etc.

RKSimon reviewed (approved):
LGTM

RKSimon committed:
Merge branch 'main' into vector_shift
